File: C#/GigaChadSys/GigaChadSys-DAO/ClaseGrupalDAOImpl.cs
using GigaChadSysDBManager;
using GigaChadSysModel;
using MySql.Data.MySqlClient;

namespace GigaChadSysDAO;

public class ClaseGrupalDAOImpl : ClaseGrupalDAO
{
    public List<ClaseGrupal> ListAll()
    {
        List<ClaseGrupal> clases = new();

        string sql = @"SELECT idClase, nombreDisciplina, descripcion,
                              duracionMinutos, nivel, activo
                       FROM ClaseGrupal
                       WHERE activo = 1";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);
            using MySqlDataReader reader = comando.ExecuteReader();

            while (reader.Read())
            {
                clases.Add(MapearClaseGrupal(reader));
            }
        }
        catch (Exception ex)
        {
            throw new Exception("Error al listar clases grupales.", ex);
        }

        return clases;
    }

    public ClaseGrupal? Load(int id)
    {
        ClaseGrupal? claseGrupal = null;

        string sql = @"SELECT idClase, nombreDisciplina, descripcion,
                              duracionMinutos, nivel, activo
                       FROM ClaseGrupal
                       WHERE idClase = @idClase";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@idClase", id);

            using MySqlDataReader reader = comando.ExecuteReader();

            if (reader.Read())
            {
                claseGrupal = MapearClaseGrupal(reader);
            }
        }
        catch (Exception ex)
        {
            throw new Exception("Error al cargar clase grupal.", ex);
        }

        return claseGrupal;
    }

    public ClaseGrupal Save(ClaseGrupal claseGrupal)
    {
        claseGrupal.Activo = true;

        string sql = @"INSERT INTO ClaseGrupal
                       (nombreDisciplina, descripcion, duracionMinutos, nivel, activo)
                       VALUES
                       (@nombreDisciplina, @descripcion, @duracionMinutos, @nivel, @activo)";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@nombreDisciplina", claseGrupal.NombreDisciplina);
            comando.Parameters.AddWithValue("@descripcion", claseGrupal.Descripcion);
            comando.Parameters.AddWithValue("@duracionMinutos", claseGrupal.DuracionMinutos);
            comando.Parameters.AddWithValue("@nivel", claseGrupal.Nivel);
            comando.Parameters.AddWithValue("@activo", claseGrupal.Activo);

            comando.ExecuteNonQuery();

            claseGrupal.IdClase = (int)comando.LastInsertedId;
        }
        catch (Exception ex)
        {
            throw new Exception("Error al registrar clase grupal.", ex);
        }

        return claseGrupal;
    }

    public ClaseGrupal? Update(ClaseGrupal claseGrupal)
    {
        string sql = @"UPDATE ClaseGrupal SET
                            nombreDisciplina = @nombreDisciplina,
                            descripcion = @descripcion,
                            duracionMinutos = @duracionMinutos,
                            nivel = @nivel,
                            activo = @activo
                       WHERE idClase = @idClase";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@nombreDisciplina", claseGrupal.NombreDisciplina);
            comando.Parameters.AddWithValue("@descripcion", claseGrupal.Descripcion);
            comando.Parameters.AddWithValue("@duracionMinutos", claseGrupal.DuracionMinutos);
            comando.Parameters.AddWithValue("@nivel", claseGrupal.Nivel);
            comando.Parameters.AddWithValue("@activo", claseGrupal.Activo);
            comando.Parameters.AddWithValue("@idClase", claseGrupal.IdClase);

            int filas = comando.ExecuteNonQuery();

            return filas > 0 ? claseGrupal : null;
        }
        catch (Exception ex)
        {
            throw new Exception("Error al actualizar clase grupal.", ex);
        }
    }
    public void Remove(ClaseGrupal claseGrupal)
    {
        claseGrupal.Activo = false;

        string sql = @"UPDATE ClaseGrupal
                       SET activo = @activo
                       WHERE idClase = @idClase";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@activo", claseGrupal.Activo);
            comando.Parameters.AddWithValue("@idClase", claseGrupal.IdClase);

            comando.ExecuteNonQuery();
        }
        catch (Exception ex)
        {
            throw new Exception("Error al eliminar clase grupal.", ex);
        }
    }

    private ClaseGrupal MapearClaseGrupal(MySqlDataReader reader)
    {
        return new ClaseGrupal
        {
            IdClase = reader.GetInt32("idClase"),
            NombreDisciplina = reader.GetString("nombreDisciplina"),
            Descripcion = reader.IsDBNull(reader.GetOrdinal("descripcion")) ? "" : reader.GetString("descripcion"),
            DuracionMinutos = reader.GetInt32("duracionMinutos"),
            Nivel = reader.IsDBNull(reader.GetOrdinal("nivel")) ? "" : reader.GetString("nivel"),
            Activo = reader.GetBoolean("activo")
        };
    }
}

File: Java/GigaChadSys-DAO/src/main/java/pe/edu/pucp/gigachadsys/dao/impl/clases/ClaseGrupalDAOImpl.java
package pe.edu.pucp.gigachadsys.dao.impl.clases;

import pe.edu.pucp.gigachadsys.dao.manager.DBManager;
import pe.edu.pucp.gigachadsys.dao.inter.clases.ClaseGrupalDAO;
import pe.edu.pucp.gigachadsys.model.clases.ClaseGrupal;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class ClaseGrupalDAOImpl implements ClaseGrupalDAO {

    @Override
    public List<ClaseGrupal> listAll() {
        List<ClaseGrupal> lista = new ArrayList<>();

        String sql =
                "SELECT idClase, nombreDisciplina, descripcion, duracionMinutos, nivel, activo " +
                        "FROM ClaseGrupal";

        try (
                Connection con = DBManager.getInstance().getConnection();
                Statement st = con.createStatement();
                ResultSet rs = st.executeQuery(sql)
        ) {
            while (rs.next()) {
                lista.add(mapearClaseGrupal(rs));
            }

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return lista;
    }

    @Override
    public ClaseGrupal load(Integer id) {
        String sql =
                "SELECT idClase, nombreDisciplina, descripcion, duracionMinutos, nivel, activo " +
                        "FROM ClaseGrupal " +
                        "WHERE idClase = ?";

        try (
                Connection con = DBManager.getInstance().getConnection();
                PreparedStatement ps = con.prepareStatement(sql)
        ) {
            ps.setInt(1, id);

            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    return mapearClaseGrupal(rs);
                }
            }

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return null;
    }

    @Override
    public ClaseGrupal save(ClaseGrupal c) {
        String sql =
                "INSERT INTO ClaseGrupal(nombreDisciplina, descripcion, duracionMinutos, nivel, activo) " +
                        "VALUES (?, ?, ?, ?, ?)";

        try (
                Connection con = DBManager.getInstance().getConnection();
                PreparedStatement ps = con.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)
        ) {
            ps.setString(1, c.getNombreDisciplina());
            ps.setString(2, c.getDescripcion());
            ps.setInt(3, c.getDuracionMinutos());
            ps.setString(4, c.getNivel());
            ps.setBoolean(5, c.getActivo());

            int affectedRows = ps.executeUpdate();

            if (affectedRows > 0) {
                try (ResultSet generatedKeys = ps.getGeneratedKeys()) {
                    if (generatedKeys.next()) {
                        c.setIdClase(generatedKeys.getInt(1));
                    }
                }
            }

            return c;

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public ClaseGrupal update(ClaseGrupal c) {
        String sql =
                "UPDATE ClaseGrupal SET " +
                        "nombreDisciplina = ?, " +
                        "descripcion = ?, " +
                        "duracionMinutos = ?, " +
                        "nivel = ?, " +
                        "activo = ? " +
                        "WHERE idClase = ?";

        try (
                Connection con = DBManager.getInstance().getConnection();
                PreparedStatement ps = con.prepareStatement(sql)
        ) {
            ps.setString(1, c.getNombreDisciplina());
            ps.setString(2, c.getDescripcion());
            ps.setInt(3, c.getDuracionMinutos());
            ps.setString(4, c.getNivel());
            ps.setBoolean(5, c.getActivo());
            ps.setInt(6, c.getIdClase());

            ps.executeUpdate();

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return c;
    }

    @Override
    public void remove(ClaseGrupal c) {
        String sql = "UPDATE ClaseGrupal SET activo = 0 WHERE idClase = ?";

        try (
                Connection con = DBManager.getInstance().getConnection();
                PreparedStatement ps = con.prepareStatement(sql)
        ) {
            ps.setInt(1, c.getIdClase());
            ps.executeUpdate();

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private ClaseGrupal mapearClaseGrupal(ResultSet rs) throws SQLException {
        ClaseGrupal c = new ClaseGrupal(
                rs.getInt("idClase"),
                rs.getString("nombreDisciplina"),
                rs.getString("descripcion"),
                rs.getInt("duracionMinutos"),
                rs.getString("nivel")
        );

        c.setActivo(rs.getBoolean("activo"));

        return c;
    }
}